src/csharp/Microsoft.Spark.Worker/Utils/AssemblyLoaderHelper.cs

```diff
@@ -0,0 +1,94 @@
+        {
+#if NETCOREAPP
+            AssemblyLoader.LoadFromFile = AssemblyLoadContext.Default.LoadFromAssemblyPath;
+            AssemblyLoadContext.Default.Resolving += (assemblyLoadContext, assemblyName) =>
+                AssemblyLoader.ResolveAssembly(assemblyName.FullName);
+#else
+            AppDomain.CurrentDomain.AssemblyResolve += (object sender, ResolveEventArgs args) =>
+                AssemblyLoader.ResolveAssembly(args.Name);
+#endif
+        }
+
+        /// <summary>
+        /// In a dotnet-interactive REPL session (driver), nuget dependencies will be
+        /// systematically added using <see cref="SparkContext.AddFile(string, bool)"/>.
+        ///
+        /// These files include:
+        /// - "{packagename}.{version}.nupkg"
+        ///   The nuget packages
+        /// - <see cref="DependencyProviderUtils.CreateFileName(ulong)"/>
+        ///   Serialized <see cref="DependencyProviderUtils.Metadata"/> object.
+        ///
+        /// On the Worker, in order to resolve the nuget dependencies referenced by
+        /// the dotnet-interactive session, we instantiate a
+        /// <see cref="DotNet.DependencyManager.DependencyProvider"/>.
+        /// This provider will register an event handler to the Assembly Load Resolving event.
+        /// By using <see cref="SparkFiles.GetRootDirectory"/>, we can access the
+        /// required files added to the <see cref="SparkContext"/>.
+        ///
+        /// Note: Because <see cref="SparkContext.AddFile(string, bool)"/> prevents
+        /// overwriting/deleting files once they have been added to the
+        /// <see cref="SparkContext"/>, numbered identifiers are added to relevant files:
+        /// - <see cref="DependencyProviderUtils.CreateFileName(ulong)"/>
+        /// </summary>
+        /// <param name="stageId">The current Stage ID</param>
+        internal static void RegisterAssemblyHandler(int stageId)
+        {
+            if (!s_runningREPL.Value || (stageId == s_stageId))
+            {
+                return;
+            }
+
+            // For a given stage, it is sufficient to instantiate one DependencyProvider.
+            // However, the Worker process may be reused between stages. New nuget dependencies
+            // may be introduced between stages and a new DependencyProvider will need to be
+            // created that can resolve them.
+            lock (s_lock)
+            {
+                if (stageId == s_stageId)
+                {
+                    return;
+                }
+                s_stageId = stageId;
+
+                var dependencyProvider = new DependencyProvider(
+                    SparkFiles.GetRootDirectory(),
+                    Directory.GetCurrentDirectory());
+                if (dependencyProvider.TryLoad())
+                {
+                    s_dependencyProvider?.Dispose();
+                    s_dependencyProvider = dependencyProvider;
+                }
```

Comment:
Fails as in returns false or throws an exception ?

Reply:
If it returns false since we are not catching any exceptions here anyway.

Reply:
It will only return false if there are no metadata files (ie, no nuget dependencies or if spark did not distribute the files to the worker) and if the cached `s_lastFileRead` matches the latest metadata file in the SparkFiles path.  In either case I don't think we need logs.

Reply:
> ie, no nuget dependencies or if spark did not distribute the files to the worker

This sounds like a critical error case, no?

Reply:
How do we determine if it's the former or latter ?

Reply:
It depends on how you implement `TryLoad`. If you think it's a critical error, returning false in that case may not work.